File: 1424.cpp
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& nums) {
        ios_base::sync_with_stdio(0);
        cin.tie(NULL);
        cout.tie(NULL);

        if (nums.size() == 1) {
            return nums[0];
        }

        queue<pair<int, int>> bfs_queue;
        bfs_queue.push({0, 0});
        vector<int> finalList;

        while(!bfs_queue.empty()) {
            pair<int, int> p = bfs_queue.front();
            bfs_queue.pop();

            finalList.push_back(nums[p.first][p.second]);

            if (p.second == 0 && p.first + 1 < nums.size()) {
                bfs_queue.push({p.first + 1, p.second});
            }

            if (p.second + 1 < nums[p.first].size()) {
                bfs_queue.push({p.first, p.second + 1});
            }
        }

        return finalList;

        // map<int, vector<int>> diagonalSumMap;

        // for(int i = 0; i < nums.size(); i++) {
        //     for(int j = 0; j < nums[i].size(); j++) {
        //         diagonalSumMap[i + j].push_back(nums[i][j]);
        //     }
        // }

        // vector<int> finalList;

        // for(auto v : diagonalSumMap) {
        //     for (int j = v.second.size() - 1; j >= 0; j--) {
        //         finalList.push_back(v.second[j]);
        //     }
        // }

        // return finalList;
    }
};

File: 1424.py
class Solution:
    def findDiagonalOrder(self, nums):
        if (len(nums) == 1):
            return nums[0]
        
        n = len(nums)
        for i in range(len(nums)):
            n = max(n, len(nums[i]))

        sumDict = [[] for _ in range(2*n - 1)]
        
        for i in range(len(nums)):
            for j in range(len(nums[i])):
                sumDict[i+j] = [nums[i][j]] + sumDict[i+j]

        finalList = []
        for i in range(len(sumDict)):
            finalList += sumDict[i]

        return finalList

# class Solution:
#     def findDiagonalOrder(self, nums):
#         if (len(nums) == 1):
#             return nums[0]
        
#         n = len(nums)
#         for i in range(len(nums)):
#             n = max(n, len(nums[i]))

#         k = 1
#         finalList = [nums[0][0]]

#         while k != n:
#             k += 1
#             i = k - 1
#             j = 0
#             while i != -1 and j != k:
#                 try:
#                     finalList.append(nums[i][j])
#                     # print(k, (i, j), nums[i][j])
#                 except:
#                     continue
#                 finally:
#                     i -= 1
#                     j += 1
        
#         temp = k - 1
#         while (k != 2):
#             k -= 1
#             i = temp
#             j = n - k
#             while j != k + n - 1:
#                 try:
#                     finalList.append(nums[i][j])
#                     # print(k, (i, j), nums[i][j])
#                 except:
#                     continue
#                 finally:
#                     i -= 1
#                     j += 1
                

#         if len(nums) == n and len(nums[-1]) == n:
#             finalList.append(nums[n-1][n-1])

#         return finalList